src/python/library/tritonclient/utils/cuda_shared_memory/__init__.py

```diff
@@ -25,117 +25,79 @@
-_ccudashm_shared_memory_region_create.restype = c_int
-_ccudashm_shared_memory_region_create.argtypes = [
-    _utf8,
-    c_uint64,
-    c_uint64,
-    POINTER(c_void_p),
-]
-_ccudashm_get_raw_handle = _ccudashm.CudaSharedMemoryGetRawHandle
-_ccudashm_get_raw_handle.restype = c_int
-_ccudashm_get_raw_handle.argtypes = [c_void_p, POINTER(c_char_p)]
-_ccudashm_shared_memory_region_set = _ccudashm.CudaSharedMemoryRegionSet
-_ccudashm_shared_memory_region_set.restype = c_int
-_ccudashm_shared_memory_region_set.argtypes = [
-    c_void_p,
-    c_uint64,
-    c_uint64,
-    c_void_p,
-    c_int,
-]
-_cshm_get_shared_memory_handle_info = _ccudashm.GetCudaSharedMemoryHandleInfo
-_cshm_get_shared_memory_handle_info.restype = c_int
-_cshm_get_shared_memory_handle_info.argtypes = [
-    c_void_p,
-    POINTER(c_void_p),
-    POINTER(c_uint64),
-    POINTER(c_uint64),
-    POINTER(c_int),
-]
-
-_cshm_cuda_shared_memory_allocate_and_read_to_host_buffer = (
-    _ccudashm.CudaSharedMemoryAllocateAndReadToHostBuffer
-)
-_cshm_cuda_shared_memory_allocate_and_read_to_host_buffer.restype = c_int
-_cshm_cuda_shared_memory_allocate_and_read_to_host_buffer.argtypes = [
-    c_void_p,
-    POINTER(c_char_p),
-]
-_cshm_cuda_shared_memory_release_host_buffer = (
-    _ccudashm.CudaSharedMemoryReleaseHostBuffer
-)
-_cshm_cuda_shared_memory_release_host_buffer.restype = c_int
-_cshm_cuda_shared_memory_release_host_buffer.argtypes = [c_char_p]
-
-_ccudashm_shared_memory_region_destroy = _ccudashm.CudaSharedMemoryRegionDestroy
-_ccudashm_shared_memory_region_destroy.restype = c_int
-_ccudashm_shared_memory_region_destroy.argtypes = [c_void_p]
-
-_ccudashm_stream_create = _ccudashm.CudaStreamCreate
-_ccudashm_stream_create.restype = c_int
-_ccudashm_stream_create.argtypes = [POINTER(c_void_p)]
-
-_ccudashm_stream_destroy = _ccudashm.CudaStreamDestroy
-_ccudashm_stream_destroy.restype = c_int
-_ccudashm_stream_destroy.argtypes = [c_void_p]
-
-_ccudashm_stream_synchronize = _ccudashm.CudaStreamSynchronize
-_ccudashm_stream_synchronize.restype = c_int
-_ccudashm_stream_synchronize.argtypes = [c_void_p]
 
 allocated_shm_regions = []
 # Internally managed stream for properly synchronizing on DLPack data,
 # the stream will be created / destroyed according to 'allocated_shm_regions'
 # and be reused throughout the process. May revisit for stream pool if
 # asynchronous write on CUDA shared memory region is requested
-_dlpack_stream = c_void_p()
+_dlpack_stream = None
 
 
-def _raise_if_error(errno):
-    """
-    Raise CudaSharedMemoryException if 'err' is non-success.
-    Otherwise return nothing.
-    """
-    if errno.value != 0:
-        ex = CudaSharedMemoryException(errno)
-        raise ex
-    return
+# Helper function to retrieve internally managed CUDA stream
+def _get_or_create_global_cuda_stream(device_id):
+    global _dlpack_stream
+    if _dlpack_stream is None:
+        _dlpack_stream = CudaStream(device_id)
+    return _dlpack_stream._stream
```

Comment:
Let me know if I am missing something, but I have a feeling that current logic will create stream on provided device only once. If we want to create another stream on another device, this will not happen. Do we consider this situation for the client? or this is not the case.

Reply:
good catch